finality-aleph/src/new_network/service.rs

```diff
@@ -165,22 +174,486 @@ impl<N: Network, D: Data> Service<N, D> {
                     }
                 },
-                Some(maybe_ready) = async {
-                    match &maybe_sender {
-                        Some(sender) => Some(sender.ready().await),
-                        None => None,
-                    }
-                } => {
-                    match self.to_send.pop_front() {
-                        Some((data, peer, _)) => {
-                            if maybe_ready.ok().map(|ready| ready.send(data.encode()).ok()).flatten().is_none() {
-                                debug!(target: "aleph-network", "Failed sending data to peer {:?}", peer);
-                            }
-                        },
-                        None => warn!(target: "aleph-network", "Attempted to send data despite empty queue."),
-                    }
-                },
+                Some(_) =  Self::send(&self.network, &mut self.to_send) => { },
             }
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::{ConnectionCommand, DataCommand, Service, IO};
+    use crate::new_network::{
+        manager::testing::MockNetworkIdentity, mock::MockNetwork, NetworkIdentity,
+        NetworkSendError, Protocol, ALEPH_PROTOCOL_NAME, ALEPH_VALIDATOR_PROTOCOL_NAME,
+    };
+    use codec::Encode;
+    use futures::{
+        channel::{mpsc, oneshot},
+        Future, StreamExt,
+    };
+    use sc_network::{
+        multiaddr::Protocol as ScProtocol, Event, Multiaddr as ScMultiaddr, ObservedRole,
+    };
+    use std::{borrow::Cow, collections::HashSet, iter, iter::FromIterator};
+    use substrate_test_runtime_client::runtime::Block;
+
+    type MockData = Vec<u8>;
+
+    pub struct MockIO {
+        messages_for_user: mpsc::UnboundedSender<(MockData, DataCommand)>,
+        messages_from_user: mpsc::UnboundedReceiver<MockData>,
+        commands_for_manager: mpsc::UnboundedSender<ConnectionCommand>,
+    }
+
+    async fn prepare() -> (
+        impl Future<Output = sc_service::Result<(), tokio::task::JoinError>>,
+        oneshot::Sender<()>,
+        MockNetwork<Block>,
+        MockIO,
+    ) {
+        let (mock_messages_for_user, messages_from_user) = mpsc::unbounded();
+        let (messages_for_user, mock_messages_from_user) = mpsc::unbounded();
+        let (mock_commands_for_manager, commands_from_manager) = mpsc::unbounded();
+        let io = IO {
+            messages_from_user,
+            messages_for_user,
+            commands_from_manager,
+        };
+        let mock_io = MockIO {
+            messages_for_user: mock_messages_for_user,
+            messages_from_user: mock_messages_from_user,
+            commands_for_manager: mock_commands_for_manager,
+        };
+
+        let (event_stream_oneshot_tx, event_stream_oneshot_rx) = oneshot::channel();
+        let network = MockNetwork::new(event_stream_oneshot_tx);
+        let service = Service::new(network.clone(), io);
+
+        let (exit_tx, exit_rx) = oneshot::channel();
+        let task_handle = async move {
```

Comment:
I would suggest rather than `run`ning the service test the methods directly. Preferably this would include rewriting the whole thing into a separate struct that just returns the correct things without sending them to any channels (like `new_network::manager::Service`), but that's probably too much of a refactor for this PR, so just using the internal methods is fine. As an additional perk this sidesteps the issue with the whole `event_stream_taken` channel, as event streams wouldn't be directly touched by these tests.

Reply:
I think that would be a huge refactor and I would prefer to keep it as it is